cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,153 @@
+GEMFILE_LOCK = "Gemfile.lock"
+
+
+class _GemMetadata(pydantic.BaseModel):
+    """
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(_GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this gem dependency."""
+        return NotImplemented
+
+
+class GitDependency(_GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        parsed_url = urlparse(url)
+        if parsed_url.scheme != "https":
+            raise ValueError("GIT dependencies must use HTTPS protocol")
+
+        return url
+
+    @pydantic.field_validator("revision")
+    def validate_git_revision(cls, revision: str) -> str:
+        """Validate that the git revision is a valid git ref."""
+        if not re.match(r"^[a-fA-F0-9]{40}$", revision):
+            raise ValueError("GIT dependencies must have a valid revision")
+
+        return revision
+
+
+class PathDependency(_GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this path dependency."""
+        return NotImplemented
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+ParseResult = list[BundlerDependency]
+
+
+def parse_lockfile(package_dir: RootedPath) -> ParseResult:
+    """Parse a Gemfile.lock file and return a list of dependencies."""
+    lockfile_path = package_dir.join_within_root(GEMFILE_LOCK)
+    gemfile_path = package_dir.join_within_root(GEMFILE)
+    if not lockfile_path.path.exists() or not gemfile_path.path.exists():
+        reason = "Gemfile and Gemfile.lock must be present in the package directory"
+        solution = (
+            "Run `bundle init` to generate the Gemfile.\n"
+            "Run `bundle lock` to generate the Gemfile.lock."
+        )
+        raise PackageRejected(reason=reason, solution=solution)
+
+    bundler_package = files("cachi2.core.package_managers.bundler")
+    lockfile_parser_script = str(bundler_package.joinpath("scripts").joinpath("lockfile_parser.rb"))
```

Comment:
This line is unnecessarily long. I am biased to actually question both the existence and usage of  `joinpath` since it returns a new `Path` instance anyway. There are other ways to write it shorter, e.g.:

- `bundler_package / "scripts/lockfile_parser.rb"`
- `Path(bundler_package, "scripts/lockfile_parser.rb")` or (if you want to stick to it) even
- `bundler_package.joinpath("scripts/lockfile_parser.rb")`

Since pathlib allows this flexibility for path concatenation we should agree on a particular style and stick to it.

Reply:
How about this ?
```
  bundler_package_path = Path(__file__).parent
  lockfile_parser_script = bundler_package_path / "scripts" / "lockfile_parser.rb"
```

Reply:
> How about this ?
> 
> ```
>     bundler_package_dir = Path(__file__).parent

Not opposing, so you can definitely do that. I like the current approach though because if we move the scripts to a different location and group them all centrally together, one only needs to tweak portion of the package location identifier with this we'd have to ditch the path. That said, unit tests would fail in such case either way, so while I do have a slight preference, ^this proposal makes perfect sense too.

>     lockfile_parser_script = bundler_package_dir / "scripts" / "lockfile_parser.rb"

why do we need to split those 2 string literals instead of just writing "scripts/lockfile_parser.rb"?

Reply:
I changed it to `scripts_dir,` so the name of the variable won't need to be changed.